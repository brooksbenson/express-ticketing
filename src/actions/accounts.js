import db from '../firebase/firebase';

export const addAccount = ({ key, ...account }) => ({
  type: 'ADD_ACCOUNT',
  key,
  account
});

export const startAddAccount = account => {
  return async dispatch => {
    const { key } = await db.ref('account_data').push(account);
    dispatch(addAccount({ key, ...account }));
    return key;
  };
};

export const setAccounts = accounts => ({
  type: 'SET_ACCOUNTS',
  accounts
});

export const startSetAccounts = () => {
  return async dispatch => {
    const snap = await db.ref('account_data').once('value');
    dispatch(setAccounts(snap.val()));
  };
};

export const updateAccount = ({ key, ...update }) => ({
  type: 'UPDATE_ACCOUNT',
  key,
  update
});

export const startUpdateAccount = ({ key, ...update }) => {
  return async dispatch => {
    await db.ref(`account_data/${key}`).set(update);
    dispatch(updateAccount({ key, ...update }));
  };
};
