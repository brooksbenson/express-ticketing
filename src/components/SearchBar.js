import React from 'react';

export default class SearchBar extends React.Component {

  state = {
    searchResults: [],
    searchString: '',
  }

  // is you can't search, text input should be empty
  static getDerivedStateFromProps(nextProps, prevState) {
    return nextProps.canSearch
      ? { ...prevState }
      : { ...prevState, searchString: '' };
  }

  displayResults() {
    return (
      <ul>
        {
          this.state.searchResults.slice(0, 7).map(r => (
            <li key={r.id} onClick={() => this.onClickHandler(r) }>
              { r[this.props.valueDisplayKey] } 
            </li>
          ))
        }
      </ul>
    )
  }

  onKeyPress = (e) => {
    if (e.key == 'Enter') {
      const [ selectedResult ] = this.state.searchResults;
      this.props.onValueSelect(selectedResult);
      this.setState(() => ({
        displaySearchResults: false,
        searchResults: [],
        searchString: selectedResult[this.props.valueDisplayKey]
      }));
    }
  };

  onResultClick = (result) => {
    this.props.onValueSelect(result);
    this.setState(() => ({
      searchString: result[this.props.valueDisplayKey],
      searchResults: []
    }));
  };

  onSearch = (e) => {
    this.props.onValueSelect(null);
    const searchString = e.target.value;
    this.setState(() => ({ 
      searchString,
      searchResults: this.props.selector(
        this.props.values,
        searchString
      )
    }));
  };

  render() {
    return (
      <div className={this.props.className}>
        <input
          autoComplete="off"
          autoFocus={this.props.canSearch}
          className="input"
          disabled={!this.props.canSearch}
          onChange={this.onSearch}
          onKeyPress={this.onKeyPress}
          placeholder={this.props.placeholder}
          ref={input => this.input = input}
          type="text"
          value={this.state.searchString}
        />
        { this.props.canSearch && this.displayResults() }
      </div>
    )
  }
}