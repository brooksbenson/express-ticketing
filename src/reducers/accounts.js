export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_ACCOUNT':
      return [...state, action.account];
    case 'SET_ACCOUNTS':
      return action.accounts;
    default:
      return state;
  }
};
