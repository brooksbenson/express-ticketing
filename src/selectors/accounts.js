export default (accounts, search) => {
  const regex = new RegExp(search.split('').join('.*?'), 'i');
  return accounts.filter(a => regex.test(a.name));
};
