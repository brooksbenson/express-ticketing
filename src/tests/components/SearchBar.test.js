import React from 'react';
import { shallow } from 'enzyme';
import SearchBar from '../../components/SearchBar';
import accountSelector from '../../selectors/accounts';
import accounts from '../../fixtures/accounts';

let wrapper, onValueSelect;
beforeEach(() => {
  onValueSelect = jest.fn();
  wrapper = shallow(
    <SearchBar
      className="search-bar"
      onValueSelect={onValueSelect}
      placeholder="Search accounts..."
      selector={accountSelector}
      values={accounts}
      valueDisplayProp="name"
    />
  );
});

test('should render SearchBar correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should set search value on input change', () => {
  const value = 'test';
  wrapper.find('input').simulate('change', { target: { value } });
  expect(wrapper.state('searchString')).toBe(value);
});

test('should render search results on input change', () => {
  wrapper.simulate('change', { target: { value: 'a' } });
  expect(wrapper.children('ul')).toBeTruthy();
});

test('should not render search results upon mounting', () => {
  expect(wrapper.children('ul')).toBeFalsy();
});

test('should set searchResults correctly', () => {
  wrapper.simulate('change', { target: { value: 'walmart' } });
  expect(wrapper.state('searchResults').length).toBe(1);
});
