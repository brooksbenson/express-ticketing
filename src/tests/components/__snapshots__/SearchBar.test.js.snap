// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchBar correctly 1`] = `
<div
  className="search-bar"
>
  <input
    autoComplete="off"
    className="input"
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Search accounts..."
    type="text"
    value=""
  />
</div>
`;
